mountpoint-s3-client/src/s3_crt_client/copy_object.rs

```diff
@@ -0,0 +1,91 @@
+use std::ops::Deref;
+use std::os::unix::prelude::OsStrExt;
+
+use mountpoint_s3_crt::{http::request_response::Header, s3::client::MetaRequestResult};
+
+use crate::object_client::{CopyObjectError, CopyObjectResult, ObjectClientResult};
+use crate::s3_crt_client::{S3CrtClient, S3Operation, S3RequestError};
+
+impl S3CrtClient {
+    /// Create and begin a new CopyObject request.
+    pub(super) async fn copy_object(
+        &self,
+        source_bucket: &str,
+        source_key: &str,
+        destination_bucket: &str,
+        destination_key: &str,
+    ) -> ObjectClientResult<CopyObjectResult, CopyObjectError, S3RequestError> {
+        // Scope the endpoint, message, etc. since otherwise rustc thinks we use Message across the await.
+        let request = {
+            let mut message = self
+                .inner
+                .new_request_template("PUT", destination_bucket)
+                .map_err(S3RequestError::construction_failure)?;
+            message
+                .set_request_path(format!("/{destination_key}"))
+                .map_err(S3RequestError::construction_failure)?;
+            message
+                .set_header(&Header::new(
+                    "x-amz-copy-source",
+                    format!("/{source_bucket}/{source_key}"),
+                ))
+                .map_err(S3RequestError::construction_failure)?;
+
+            let span = request_span!(
+                self.inner,
+                "copy_object",
+                source_bucket,
+                source_key,
+                destination_bucket,
+                destination_key
+            );
+
+            self.inner
+                .make_simple_http_request(message, S3Operation::CopyObject, span, parse_copy_object_error)?
+        };
+
+        let _body = request.await?;
+
+        Ok(CopyObjectResult {})
+    }
+}
+fn parse_copy_object_error(result: &MetaRequestResult) -> Option<CopyObjectError> {
+    match result.response_status {
+        403 => {
```

Comment:
How do we decide that we need to parse only 403:ObjectNotInActiveTierError?

Reply:
This is from the official documentation for copy obj api https://docs.aws.amazon.com/AmazonS3/latest/API/API_CopyObject.html#API_CopyObject_Errors